### PR TITLE
[c-libcurl]: add support for url encoding

### PR DESCRIPTION
**Pull request by zhemant**
Added support for url encoding (present only in the query params)

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [ ] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@wing328  @ityuhui

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
Related to this, I have opened https://github.com/OAI/OpenAPI-Specification/issues/2119 and #5117. I think it would be great to address URL path and query encoding in a consistent way across languages.

**Comment by zhemant** [comment]
For URL path encoding, I believe it should have support right from the library instead of reinventing the whole encoding wheel again. I agree with the example you have given. But with libcurl I tried and it encodes "/" also this creates the problem in the server where the handlers do not recognize the URLs as they are encoded. Let me know if you get any update on encodings.

**Comment by sebastien-rosset** [comment]
> For URL path encoding, I believe it should have support right from the library instead of reinventing the whole encoding wheel again.

Yes, I completely agree we should not reinvent the wheel. By library, do you mean codegen? Per-language encoding under modules/openapi-generator/src/main/java/org/openapitools/codegen/languages? A mustache template invoking a language specific library? I think we may need a combination, but ideally as much as possible would be pushed to codegen.

I think codegen should be able to do some sanitization even before the 

> I agree with the example you have given. But with libcurl I tried and it encodes "/" also this creates the problem in the server where the handlers do not recognize the URLs as they are encoded. Let me know if you get any update on encodings.

Yes, I had a similar problem with golang. The standard golang library performs incorrect path encoding, and I had to use a different library that carefully encodes the URL path.

**Comment by zhemant** [comment]
By library I mean for C client, we use the curl library to send the request. Curl library provides URL encoding functionality. I tried using it for the whole path but it creates more problem.  and CURL is a standard library so my guessing is it is updated as per the required standards(what to encode and what not to encode).

File: modules/openapi-generator/src/main/resources/C-libcurl/apiClient.c.mustache [hunk 1]
```diff
@@ -181,14 +181,20 @@ char *assembleTargetUrl(char    *basePath,
...
         strcat(targetUrl, "?");
         list_ForEach(listEntry, queryParameters) {
             keyValuePair_t *pair = listEntry->data;
             replaceSpaceWithPlus(pair->key);
             strcat(targetUrl, pair->key);
             strcat(targetUrl, "=");
-            replaceSpaceWithPlus(pair->value);
-            strcat(targetUrl, pair->value);
+            if(curl) {
+              char *output = curl_easy_escape(curl, pair->value, 0);
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/resources/C-libcurl/apiClient.c.mustache]: PATH   : Http://SomeUrl.com:8080//store/curlybraces-%7B%7D/AZaz-09._~!$&'(abc)*+, ;=:@Œë🍇ℬ/order/template-{orderId}/.././c///g?c=3&a=1&b=9&c=0#target

This is how the above URL is supposed to be converted. Note characters like :, /, $ ! are not percent encoded.

ESCAPED: http://someurl.com:8080//store/curlybraces-%7B%7D/AZaz-09._~!$&'(abc)*+,%20;=:@%C5%92%C3%AB%F0%9F%8D%87%E2%84%AC/order/template-{orderId}/.././c///g?c=3&a=1&b=9&c=0#target

Using curl_easy_escape, I get a lot more character escapes:
Http%3A%2F%2FSomeUrl.com%3A8080%2F%2Fstore%2Fcurlybraces-%257B%257D%2FAZaz-09._~%21%24%26%27%28abc%29%2A%2B%2C%20%3B%3D%3A%40%C5%92%C3%AB%F0%9F%8D%87%E2%84%AC%2Forder%2Ftemplate-%7BorderId%7D%2F..%2F.%2Fc%2F%2F%2Fg%3Fc%3D3%26a%3D1%26b%3D9%26c%3D0%23target

**Pull request closed without merging**